lib/pure/hashes.nim

```diff
@@ -136,131 +139,146 @@ proc hash*[T: Ordinal](x: T): Hash {.inline.} =
   ## Efficient hashing of other ordinal types (e.g. enums).
   result = ord(x)
 
+template bytewiseHashing(result: Hash, x: typed, start, stop: int) =
+  for i in start .. stop:
+    result = result !& hash(x[i])
+  result = !$result
+
+template hashImpl(result: Hash, x: typed, start, stop: int) =
+  let
+    elementSize = sizeof(x[start])
+    stepSize = IntSize div elementSize
+  var i = start
+  while i <= stop+1 - stepSize:
+    var n = 0
+    when nimvm:
+      # we cannot cast in VM, so we do it manually
+      for j in countdown(stepsize-1, 0):
```

Comment:
This trick can also be used to implement ``hashIgnoreStyle`` efficiently, or ``hashIgnoreStyle`` (same for hashIgnoreCase) could use the old byte-wise method. strtabs uses these and quite some code uses strtabs so I consider it important to not slow these down further.

Reply:
> This trick can also be used to implement `hashIgnoreStyle` efficiently

Not really, since we don't know the amount of underscores in the string.

> I consider it important to not slow these down further.

I've run a benchmark similar to the one above where I create random strings of different lengths with random amount of underscores and with both upper- and lower-cased letters (e.g.: `Bx_f`, `_vKzBtMa`, `FjAmW__sVdEsWbGi___r_zL_MeKvW_G` and so on) and then I run `hashIgnoreStyle`.

The results of that benchmark show that on average there is no significant difference in speed between the old and new algorithm. Sometimes the old one is slightly faster, sometimes the new one, but the differences are under 2% (compare that with up to 320% improvements for `hash(string)`).